### ISSUE TITLE
Forced slowdown in Legend of Legaia hyper arts, items and magic?

### ISSUE THREAD

**Issue by Rumba121** [libretro/pcsx_rearmed#515]
The issue also occurs in standalone PCSX-ReARMed, even with unai renderer.

## Description

Its not a major issue, yet an annoying one. In this game (scus version) while in combat you can execute hyper arts, spirit and magic summoning all appear to have an initial slowdown which desyncs the audio from the image.
¿any idea how to solve this?

## Steps to reproduce

Play the game with any renderer open-gl or dispmanx or unai, enter a combat and execute a "Hyper Art", use an Item, cast a Magic or use Spirit.

## When did the behavior start?

Always.

## Your device/OS/platform/architecture

Device pi zero W, pi 3 a+ and pi 3 b+
OS raspbian buster with retroarch 1.9.1 with Lakka nb (from 21-03-30) cores, Lakka nb and retropie 4.7.1

## Logs (enable file logging and set log levels to DEBUG for core and frontend)
[retroarch__2021_04_23__17_57_43.log](https://github.com/libretro/pcsx_rearmed/files/6368097/retroarch__2021_04_23__17_57_43.log)

## Screenshots (if needed for visual confirmation)

No

## Others (save states and/or save files nearest to the affected area, compressed)

No

**Comment by gameblabla** [comment]
Does this still happen ? I believe this is a counting issue just like how some games suffer from this.

**Comment by JosepMC154** [comment]
This does seem to be still happening at the present. (Also somewhat-related, I can hear some noise/hissing on this game even with sound interpolation set on Gaussian, unlike in DuckStation/Beetle PSX.)

**Comment by notaz** [comment]
How to reproduce this? I went to the tutorial fight with Tetsu, which involves abilities mentioned in the report and couldn't notice anything wrong. What device and which version are you using?

**Comment by JosepMC154** [comment]
"r24l 9fec7ce" on PC/Xbox Series S and "r24l 92bb759d" on RG353V with PSP BIOS, NTSC version with [Restored Progression hack](https://www.romhacking.net/hacks/8080/). It's a bit hard to tell on its own because one can simply assume it's the game itself, but if you compare with [footage on real hardware](https://www.youtube.com/live/iiE2lnw5uEI?si=CsJcSSZNJ_dwD4Bi&t=215), you can see there's no slowdown in things like hyper arts.

Here's a save about 6-7 hours in, with some arts and magic unlocked to check out. Beetle PSX behaves more like in real hardware.

[Legend of Legaia (USA) - Restored Progression (v1.0).zip](https://github.com/user-attachments/files/18443095/Legend.of.Legaia.USA.-.Restored.Progression.v1.0.zip)

**Comment by notaz** [comment]
I don't know what you're getting but I fail to see much difference from the footage.

Could you try boosting "PSX CPU clock" in "Core Options -> System"? If it helps I'd like to know the minimum value that gives closest speed to what it should be.

**Comment by JosepMC154** [comment]
I don't know what to say, then. For me it seems to happen especially when using hyper arts like Tornado Flame or Frost Breath, where the character animation slows down a little and the audio desyncs momentarily. Maybe it'd be better to ask for more feedback?

Boosting the CPU clock makes everything smoother, but doesn't quite fix it...

If at least this helps, the audio noise/hissing I was hearing is gone after turning off the SPU's reverb, so that seems to be the culprit there.

**Comment by notaz** [comment]
I can't seem to hear any hiss either. Here is an audio recording on how it sounds here doing Tornado Flame, does this have the hiss? I guess the slight choppiness during "explosions" is the first issue?

https://github.com/user-attachments/assets/797f6333-7775-4248-b162-e9e4b87956bd

**Comment by JosepMC154** [comment]
Yup, that choppiness happens because of that was stated in the OP.

About the "hiss", I meant in the game as a whole, playing the sequenced BGM and unrelated to the thread's issue. It can be heard occassionally while playing with reverb on, most noticiable while opening and closing the pause menu. My fault for going off on a related tangent.

Other than these two things, the game runs pretty well. None of them are dealbreakers, just annoyances to keep track of in hopes they can get fixed eventually. So don't feel pressured to look into it.

**Comment by notaz** [comment]
Please try the latest version.
- I've updated the reverb code. The correct formula was documented some years back, but I did not bother to update it as the old one sounded good enough for me until now.
- During the special attacks the game streams audio directly from CD for some reason, so if the game logic runs slower (due to lack of CPU time for example) the animations will start lagging behind. Audio stutter was from a different regression.

**Comment by JosepMC154** [comment]
Sorry for the late reply. I just updated and I think the gameplay issue has been tackled. However, I am not an expert of this game, so I can't be sure 100%, but it does seem to fare better now. It would be great if more people could test it out.

As for the audio, it still has white noise with SPU reverb on. I've recorded audio from the title screen with the latest builds available for both PCSX-ReARMed (2ff70f0 on Windows 10) and Beetle PSX, so that you can check the difference.

https://github.com/user-attachments/assets/ab778164-5c20-414d-a4e0-25c109bf803e 

https://github.com/user-attachments/assets/fac8fbbd-9a31-4ea4-a297-69dfa31effc6

**Comment by notaz** [comment]
While I can hear some subtle cracking-like thing in your recording, I'm unable to get this from the emu myself. I think there is some kind of post-processing going on on your system. Maybe it's RetroArch, maybe some enhancement done by your sound drivers or something. Can you attach your RetroArch log, or at least check for `[Audio]` lines in that log?

What I suspect is going on is that PCSX is doing much less audio filtering compared to Beetle (for performance reasons) leaving some (mostly?) inaudible artifacts that are boosted by some audio post-processing or extra resampling on your system.

**Comment by notaz** [comment]
I've added some filtering, try how it sounds. I can't hear any difference, but the spectrogram has much less stuff going on in the high frequencies. Also check what resampler and rate you have in RetroArch settings (Settings - Audio - Output).

**Comment by JosepMC154** [comment]
Thanks for the update, unfortunately I can still hear it... The only post-processing I have going is Equalizer APO w/ Peace Audio for EQ purposes, but turning it off doesn't change anything. Resampler is set to sinc and Output Rate to 44100 Hz.

Here's all the audio-related info I could find on my RetroArch log while running the game. (Also, SwanStation doesn't do it for me either/behaves like Beetle PSX.)

> [INFO] [Environ]: SET_AUDIO_BUFFER_STATUS_CALLBACK.
> [INFO] [Core]: Geometry: 256x240, Aspect: 1.333, FPS: 60.00, Sample rate: 44100.00 Hz.
> [INFO] [Audio]: Set audio input rate to: 44100.00 Hz.
> [INFO] [WASAPI]: Client initialized (shared, PCM, 44100Hz, 64.0ms).
> [INFO] [Audio]: Started synchronous audio driver.
> [INFO] [Environ]: RETRO_ENVIRONMENT_SET_MINIMUM_AUDIO_LATENCY.

**Comment by notaz** [comment]
Since I'm unable to do it myself, can you rate this audio? This is logged directly from the core's internals, Linux version, and would eliminate any possible bad interaction with RetroArch, settings, etc.

https://github.com/user-attachments/assets/6c7d5472-d8d0-4310-a5db-1b76c812e246

It's also really strange that you said it's most noticeable on the menu sounds when the game has reverb processing disabled on them, they are not sent to the reverb unit at all...

**Comment by JosepMC154** [comment]
This recording sounds fine, can't hear the white noise from my end. 

I mentioned the menu in particular because it seems most audible when the sound effects intersect with the BGM. While playing the game on my end, it was also particularly noticiable on things like the large volume changes during the overworld theme.

**Comment by notaz** [comment]
Then I have no idea why it does it for you then. It could do it before the fixes since the reverb code was occasionally reading a single sample out of bounds, which then could get reverberated and produce the effect from your recording. I would suggest RetroArch might be picking up an old core from somewhere for you but you confirmed other fixes so I'm at loss there.

Another unlikely possibility is libretro.com build server miscompiling the core somehow, here is my 64bit windows build:
[pcsx_rearmed_libretro.dll.zip](https://github.com/user-attachments/files/18805353/pcsx_rearmed_libretro.dll.zip)

**Comment by JosepMC154** [comment]
Haven't compared it with the most recent nightly build on RA, but although it's not completely gone, it feels much improved, and the white noise seems occurs much less frequently, often times having to pay attention to notice it. I'll try to check on other devices, but really, thank you for your efforts.

**Comment by JosepMC154** [comment]
I'm terribly sorry for the late reply and bothering you with this, but I think I found the culprit. Turns out I had runahead running on the core for fighting games, and that's what was causing the white noise I was getting. Turning it off made it sound just like in your recording. My apologies again and thank you for all your work in PCSX-ReARMed.

With that out of the way, I think OP's issue is finally resolved. So unless someone else wants to chime in, I think you can finally close this thread.

**Comment by notaz** [comment]
Oh that explains it. Run-ahead being broken indicates savestate issues, I've patched a problem with it + threaded SPU being enabled so try updating.

And yeah closing this, you can open a new issue if there is still any problem with run-ahead or otherwise.
